lib/common/ipc_pacemakerd.c

```diff
@@ -0,0 +1,249 @@
+        pcmk_pacemakerd_reply_unknown
+    };
+
+    if (private->replies_expected > 0) {
+        private->replies_expected--;
+    }
+
+    value = crm_element_value(reply, F_CRM_MSG_TYPE);
+    if ((value == NULL) || (strcmp(value, XML_ATTR_RESPONSE))) {
+        crm_debug("Unrecognizable pacemakerd message: invalid message type '%s'",
+                  crm_str(value));
+        status = CRM_EX_PROTOCOL;
+        reply = NULL;
+    }
+
+    if (crm_element_value(reply, XML_ATTR_REFERENCE) == NULL) {
+        crm_debug("Unrecognizable pacemakerd message: no reference");
+        status = CRM_EX_PROTOCOL;
+        reply = NULL;
+    }
+
+    value = crm_element_value(reply, F_CRM_TASK);
+    if ((value == NULL) || strcmp(value, CRM_OP_PING)) {
+        crm_debug("Unrecognizable pacemakerd message: '%s'", crm_str(value));
+        status = CRM_EX_PROTOCOL;
+        reply = NULL;
+    }
+
+    // Parse useful info from reply
+
+    if (reply != NULL) {
+        xmlNode *msg_data = get_message_xml(reply, F_CRM_DATA);
+        const char *status =
+            crm_element_value(msg_data, XML_PING_ATTR_STATUS);
+        long long value_ll = 0;
+
+        reply_data.reply_type = pcmk_pacemakerd_reply_ping;
+        reply_data.data.ping.state =
+            pcmk_pacemakerd_api_pacemakerd_state_text2enum(
+                crm_element_value(msg_data, XML_PING_ATTR_PACEMAKERDSTATE));
+        reply_data.data.ping.status =
+            crm_str_eq(status, "ok", FALSE)?pcmk_rc_ok:pcmk_rc_error;
+
+        crm_element_value_ll(msg_data, XML_ATTR_TSTAMP, &value_ll);
+        reply_data.data.ping.last_good = (time_t) value_ll;
+        reply_data.data.ping.sys_from = crm_element_value(msg_data,
+                                            XML_PING_ATTR_SYSFROM);
+    }
+
+    pcmk__call_ipc_callback(api, pcmk_ipc_event_reply, status, &reply_data);
+}
+
+pcmk__ipc_methods_t *
+pcmk__pacemakerd_api_methods()
+{
+    pcmk__ipc_methods_t *cmds = calloc(1, sizeof(pcmk__ipc_methods_t));
+
+    if (cmds != NULL) {
+        cmds->new_data = new_data;
+        cmds->free_data = free_data;
+        cmds->post_connect = post_connect;
+        cmds->reply_expected = reply_expected;
+        cmds->dispatch = dispatch;
+        cmds->post_disconnect = post_disconnect;
+    }
+    return cmds;
+}
+
+int
+pcmk_pacemakerd_api_ping(pcmk_ipc_api_t *api, const char *ipc_name)
+{
+    pacemakerd_api_private_t *private;
+    xmlNode *cmd;
+    int rc;
+
+    CRM_CHECK(api != NULL, return -EINVAL);
+    private = api->api_data;
+    CRM_ASSERT(private != NULL);
+
+    cmd = create_request(CRM_OP_PING, NULL, NULL, CRM_SYSTEM_MCP,
+        ipc_name?ipc_name:((crm_system_name? crm_system_name : "client")),
+        private->client_uuid);
+
+    if (cmd) {
+        rc = pcmk__send_ipc_request(api, cmd);
+        if (rc != pcmk_rc_ok) {
+            crm_debug("Couldn't ping pacemakerd: %s rc=%d",
+                pcmk_strerror(rc), rc);
+            rc = -ECOMM;
```

Comment:
the new return codes use pcmk_rc_str() instead of pcmk_strerror(), and positive system errno's

Reply:
one of the few lines left over from the previous implementation ;-)